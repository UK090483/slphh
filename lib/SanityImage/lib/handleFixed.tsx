import type { ImageMetaResult } from "../query";
import { IImagePropsWithDimensions, UseSanityImageOptions } from "../types";
import getDefaultResult from "./getDefaultResult";
import handleCrop from "./handleCrop";
import { checkDimensionOptions } from "./helper";

const handleFixed = (
  image: ImageMetaResult,
  options: UseSanityImageOptions
): IImagePropsWithDimensions => {
  const imageProps = {
    ...getDefaultResult(image),
    width: image.width,
    height: image.height,
  };

  const {
    justWidth,
    justHeight,
    widthAndHeight,
    noDimensions,
    width,
    height,
    aspectRatio,
    crop,
  } = getDimensions(image, options);

  if (justWidth) {
    imageProps.width = width;
    imageProps.height = width / aspectRatio;
  }
  if (justHeight) {
    imageProps.width = height * aspectRatio;
    imageProps.height = height;
  }
  if (widthAndHeight) {
    imageProps.width = width;
    imageProps.height = height;
  }
  if (noDimensions) {
    imageProps.width = image.width;
    imageProps.height = image.width / aspectRatio;
  }

  if (crop) {
    imageProps.src = `${imageProps.src}${crop.query}`;
  }

  return { ...options, ...imageProps };
};

export default handleFixed;

const getDimensions = (
  image: ImageMetaResult,
  options: UseSanityImageOptions
) => {
  const crop = handleCrop(image);
  const { width, height } = checkDimensionOptions(options);

  const res = {
    crop: crop,
    justWidth: width && !height,
    justHeight: !width && height,
    widthAndHeight: !!width && !!height,
    noDimensions: !width && !height,
    width: width || image.width,
    height: height || image.height,
    aspectRatio: crop ? crop.dimensions.aspectRatio : image.aspectRatio,
  };

  return res;
};
