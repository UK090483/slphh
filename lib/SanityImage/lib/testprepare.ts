import type { ImageMetaResult } from "../query";

export const imageMeta: ImageMetaResult = {
  alt: "testAlt",
  aspectRatio: 1.412909503730133,
  crop: null,
  height: 3083,
  id: "4ba76be5814da24ee8ea5dd0e6e0225c9cfccf29",
  lqip: "data:image/jpeg;base64,/9j/2wBDAAYEBQYFBAYGBQYHBwYIChAKCgkJChQODwwQFxQYGBcUFhYaHSUfGhsjHBYWICwgIyYnKSopGR8tMC0oMCUoKSj/2wBDAQcHBwoIChMKChMoGhYaKCgoKCgoKCgoKCgoKCgoKCgoKCgoKCgoKCgoKCgoKCgoKCgoKCgoKCgoKCgoKCgoKCj/wAARCAAOABQDASIAAhEBAxEB/8QAFwABAQEBAAAAAAAAAAAAAAAABwADBv/EACEQAAIBBAIDAQEAAAAAAAAAAAECAwAEBREGQRIiMSFh/8QAFQEBAQAAAAAAAAAAAAAAAAAABAX/xAAdEQABBAIDAAAAAAAAAAAAAAABAAIDBBIhMZHw/9oADAMBAAIRAxEAPwAowXHiltdJcxacjwB73/K0uOK4eJZtXU0rrEGCL9DHo098S4vj8xjRd3KnzLkMB8OqMLrGtjuX362giEDzHSts60aBQsm1IcRwNhWZREwBh7R9keIGCZBDeRFGQN7+pG+iKq7PNYVL3ISTSvpj+aHVVWhSf4opdCTpf//Z",
  type: "image/jpeg",
  url: "https://cdn.sanity.io/images/jgnu3d9f/production/4ba76be5814da24ee8ea5dd0e6e0225c9cfccf29-4356x3083.jpg",
  width: 4356,
};

export const imageMeatWithCrop = {
  ...imageMeta,
  crop: {
    _type: "sanity.imageCrop",
    bottom: 0,
    left: 0.0034722222222222225,
    right: 0.517361111111111,
    top: 0,
  },
};

export const imageMeatWithHotspot = {
  ...imageMeta,
  hotspot: {
    _type: "sanity.imageHotspot",
    height: 1,
    width: 0.47916666666666674,
    x: 0.24305555555555558,
    y: 0.5,
  },
};
