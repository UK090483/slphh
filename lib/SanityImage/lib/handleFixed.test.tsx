import handleFixed from "./handleFixed";

import { imageMeta, imageMeatWithCrop } from "./testprepare";

const width = 123;
const height = 345;

describe("handleFixed", () => {
  it("type click should render ", () => {
    const res = handleFixed(imageMeta, {});
    expect(res.width).toBe(imageMeta.width);
    expect(res.height).toBe(imageMeta.height);
  });

  it("should handle width", () => {
    const res = handleFixed(imageMeta, { width });
    expect(res.width).toBe(width);
    expect(res.height).toBe(width / imageMeta.aspectRatio);

    const resCrop = handleFixed(imageMeatWithCrop, { width });
    expect(resCrop.width).toBe(width);
    expect(resCrop.height).toMatchSnapshot();
  });

  it("should handle height", () => {
    const res = handleFixed(imageMeta, { height });
    expect(res.height).toBe(height);
    expect(res.width).toBe(height * imageMeta.aspectRatio);

    const resCrop = handleFixed(imageMeatWithCrop, { height });
    expect(resCrop.height).toBe(height);
    expect(resCrop.width).toMatchSnapshot();
  });

  it("should handle with and height", () => {
    const res = handleFixed(imageMeta, { height, width });
    expect(res.height).toBe(height);
    expect(res.width).toBe(width);
  });
});
