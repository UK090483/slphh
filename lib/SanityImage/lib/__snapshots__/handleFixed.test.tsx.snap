// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handleFixed should handle height 1`] = `233.5436263379825`;

exports[`handleFixed should handle width 1`] = `181.70052707235266`;
